### PR TITLE
MAINT: lint: enable UP rules

### PR DESCRIPTION
**Pull request by lucascolley**
#### Reference issue
https://github.com/scipy/scipy/pull/18353#issuecomment-1808711017

#### What does this implement/fix?
- The UP rules are enabled in the linter
- All existing violations are cleaned up

Quoting @stefanv :

> The [Ruff PyUpgrade rules](https://docs.astral.sh/ruff/rules/#pyupgrade-up) ensure that SciPy uses modern Python constructs that work with all support Python versions.
> While the linter only runs on files touched in any given PR, we have the option to apply these rules wholesale each time we drop support for another version of Python.
> This keeps our code current, at low cost to maintainers.

#### Additional information
The vast majority of these changes were made automatically by Ruff's `--fix`, with the rest (changing from `Union` typing and to `yield from`) made by me.

Due to the automatic fixes, further (minor) style cleanups are possible here, but I don't know if they're worth the effort (or worth delaying this PR for).

The blame ignore will be added once this is ready.

### CODE REVIEW

**Comment by stefanv** [comment]
If there is agreement on this change, we should add a commit with an entry in `.git-blame-ignore-revs`, and then merge (not squash merge) the PR.

**Comment by lucascolley** [comment]
> If there is agreement on this change, we should add a commit with an entry in .git-blame-ignore-revs, and then merge (not squash merge) the PR.

Yes, I didn't yet in anticipation of a force push (which was immediately necessary since I forgot to `mamba upgrade ruff`). Perhaps there will be no force pushes from now on but no harm in holding off for a little longer.

**Comment by lucascolley** [comment]
CI is green (apart from the broken docs build). I've noticed some formatting issues (multiline strings off by 1 space, closing brackets on a new line). These will probably have to be looked at manually unfortunately.

**Comment by ilayn** [comment]
> These will probably have to be looked at manually unfortunately.

It would difficult to catch them later if we lose the `.format()` style. Because then the tool will not complain about anything. 

I see quite a number of 

```python
    ('some long string'
    '')   # <- This looks like a tool bug to me
```

instances and I think ruff should be fixed about that.

**Comment by lucascolley** [comment]
> This looks like a tool bug to me

IIUC this is because the linter is intended to be used in tandem with a formatter. This is difficult for SciPy to adopt short of formatting the entire codebase (far too disruptive).

**Comment by melissawm** [comment]
Docs build failure is fixed in main, not related to this PR.

File: scipy/_build_utils/system_info.py [hunk 1]
```diff
@@ -30,8 +30,8 @@ def combine_dict(*dicts, **kw):
                 elif value == old_value:
                     continue
 
-                raise ValueError("Conflicting configuration dicts: {!r} {!r}"
-                                 "".format(new_dict, d))
+                raise ValueError(f"Conflicting configuration dicts: {new_dict!r} {d!r}"
+                                 "")
```
Comment (mdhaber) [inline scipy/_build_utils/system_info.py:34]: This was a bit over the ugliness tolerance threshold.
```suggestion
                message = f"Conflicting configuration dicts: {new_dict!r} {d!r}"
                raise ValueError(message)
```

File: scipy/_lib/decorator.py [hunk 2]
```diff
@@ -221,8 +221,8 @@ def decorator(caller, _func=None):
     if inspect.isclass(caller):
         name = caller.__name__.lower()
         callerfunc = get_init(caller)
-        doc = 'decorator({}) converts functions/generators into ' \
-            'factories of {} objects'.format(caller.__name__, caller.__name__)
+        doc = f'decorator({caller.__name__}) converts functions/generators into ' \
+            f'factories of {caller.__name__} objects'
```
Comment (mdhaber) [inline scipy/_lib/decorator.py:225]:
```suggestion
        doc = (f'decorator({caller.__name__}) converts functions/generators into ' 
               f'factories of {caller.__name__} objects')
```

File: scipy/constants/_codata.py [hunk 3]
```diff
@@ -1567,8 +1567,7 @@ class ConstantWarning(DeprecationWarning):
 
 def _check_obsolete(key: str) -> None:
     if key in _obsolete_constants and key not in _aliases:
-        warnings.warn("Constant '{}' is not in current {} data set".format(
-            key, _current_codata), ConstantWarning)
+        warnings.warn(f"Constant '{key}' is not in current {_current_codata} data set", ConstantWarning)
```
Comment (mdhaber) [inline scipy/constants/_codata.py:1570]: Why is this passing as-is?
```suggestion
        warnings.warn(f"Constant '{key}' is not in current {_current_codata} data set",
                      ConstantWarning)
```

File: scipy/fft/_pocketfft/basic.py [hunk 4]
```diff
@@ -23,8 +23,8 @@ def c2c(forward, x, n=None, axis=-1, norm=None, overwrite_x=False,
         tmp, copied = _fix_shape_1d(tmp, n, axis)
         overwrite_x = overwrite_x or copied
     elif tmp.shape[axis] < 1:
-        raise ValueError("invalid number of data points ({}) specified"
-                         .format(tmp.shape[axis]))
+        raise ValueError(f"invalid number of data points ({tmp.shape[axis]}) specified"
+                         )
```
Comment (mdhaber) [inline scipy/fft/_pocketfft/basic.py:27]: Ugh I guess there are several of these. I guess we can let it go.
```suggestion
        message = f"invalid number of data points ({tmp.shape[axis]}) specified"
        raise ValueError(message)
```

File: scipy/interpolate/_fitpack_impl.py [hunk 5]
```diff
@@ -744,8 +744,8 @@ def splder(tck, n=1):
     t, c, k = tck
 
     if n > k:
-        raise ValueError(("Order of derivative (n = {!r}) must be <= "
-                          "order of spline (k = {!r})").format(n, tck[2]))
+        raise ValueError(f"Order of derivative (n = {n!r}) must be <= "
+                          f"order of spline (k = {tck[2]!r})")
```
Comment (mdhaber) [inline scipy/interpolate/_fitpack_impl.py:748]:
```suggestion
        raise ValueError(f"Order of derivative (n = {n!r}) must be <= "
                         f"order of spline (k = {tck[2]!r})")
```

File: scipy/interpolate/_interpolate.py [hunk 6]
```diff
@@ -345,8 +345,7 @@ def __call__(self, x, y, dx=0, dy=0, assume_sorted=False):
             any_out_of_bounds_y = np.any(out_of_bounds_y)
 
         if self.bounds_error and (any_out_of_bounds_x or any_out_of_bounds_y):
-            raise ValueError("Values out of range; x must be in %r, y in %r"
-                             % ((self.x_min, self.x_max),
+            raise ValueError("Values out of range; x must be in {!r}, y in {!r}".format((self.x_min, self.x_max),
```
Comment (mdhaber) [inline scipy/interpolate/_interpolate.py:348]: Similar here - why is this passing the line length check?

File: scipy/io/wavfile.py [hunk 7]
```diff
@@ -658,9 +658,9 @@ def read(filename, mmap=False):
                 if data_chunk_received:
                     # End of file but data successfully read
                     warnings.warn(
-                        "Reached EOF prematurely; finished at {:d} bytes, "
-                        "expected {:d} bytes from header."
-                        .format(fid.tell(), file_size),
+                        f"Reached EOF prematurely; finished at {fid.tell():d} bytes, "
+                        f"expected {file_size:d} bytes from header."
+                        ,
```
Comment (mdhaber) [inline scipy/io/wavfile.py:663]:
```suggestion
                        f"expected {file_size:d} bytes from header.",
```

File: scipy/linalg/_decomp.py [hunk 8]
```diff
@@ -523,16 +523,16 @@ def eigh(a, b=None, *, lower=True, eigvals_only=False, overwrite_a=False,
...
-            raise ValueError('"{}" does not accept input b array '
+            raise ValueError(f'"{driver}" does not accept input b array '
                              'for standard eigenvalue problems.'
-                             ''.format(driver))
+                             '')
         if subset and (driver in ["ev", "evd", "gv", "gvd"]):
-            raise ValueError('"{}" cannot compute subsets of eigenvalues'
-                             ''.format(driver))
+            raise ValueError(f'"{driver}" cannot compute subsets of eigenvalues'
+                             '')
```
Comment (mdhaber) [inline scipy/linalg/_decomp.py:535]: And similar above.
```suggestion
            raise ValueError(f'"{driver}" cannot compute subsets of eigenvalues')
```

File: scipy/ndimage/_measurements.py [hunk 9]
```diff
@@ -519,9 +519,8 @@ def labeled_comprehension(input, labels, index, func, out_dtype, default, pass_p
 
     index = numpy.atleast_1d(index)
     if np.any(index.astype(labels.dtype).astype(index.dtype) != index):
-        raise ValueError("Cannot convert index values from <%s> to <%s> "
-                            "(labels' type) without loss of precision" %
-                            (index.dtype, labels.dtype))
+        raise ValueError(f"Cannot convert index values from <{index.dtype}> to <{labels.dtype}> "
+                            "(labels' type) without loss of precision")
```
Comment (mdhaber) [inline scipy/ndimage/_measurements.py:523]:
```suggestion
        raise ValueError(f"Cannot convert index values from <{index.dtype}> to <{labels.dtype}> "
                         "(labels' type) without loss of precision")
```

File: scipy/odr/_odrpack.py [hunk 10]
```diff
@@ -261,9 +261,8 @@ def __init__(self, x, y=None, we=None, wd=None, fix=None, meta=None):
         self.x = _conv(x)
 
         if not isinstance(self.x, numpy.ndarray):
-            raise ValueError(("Expected an 'ndarray' of data for 'x', "
-                              "but instead got data of type '{name}'").format(
-                    name=type(self.x).__name__))
+            raise ValueError("Expected an 'ndarray' of data for 'x', "
+                              f"but instead got data of type '{type(self.x).__name__}'")
```
Comment (mdhaber) [inline scipy/odr/_odrpack.py:265]:
```suggestion
                             f"but instead got data of type '{type(self.x).__name__}'")
```

File: scipy/odr/_odrpack.py [hunk 11]
```diff
@@ -375,9 +374,8 @@ def __init__(self, x, y=None, sx=None, sy=None, covx=None, covy=None,
...
```
Comment (mdhaber) [inline scipy/odr/_odrpack.py:378]:
```suggestion
            raise ValueError("Expected an 'ndarray' of data for 'x', "
                             f"but instead got data of type '{type(self.x).__name__}'")
```

File: scipy/optimize/_shgo_lib/_complex.py [hunk 12]
```diff
@@ -120,7 +120,7 @@ def __init__(self, dim, domain=None, sfield=None, sfield_args=(),
         # Domains
         self.domain = domain
         if domain is None:
-            self.bounds = [(float(0), float(1.0)), ] * dim
+            self.bounds = [(float(0), 1.0), ] * dim
```
Comment (mdhaber) [inline scipy/optimize/_shgo_lib/_complex.py:123]:
```suggestion
            self.bounds = [(0.0, 1.0), ] * dim
```

File: scipy/stats/tests/test_distributions.py [hunk 13]
```diff
@@ -5271,10 +5271,8 @@ def test_pdf_nolan_samples(
                     p,
                     subdata['p'],
                     rtol,
-                    err_msg="pdf test %s failed with method '%s'"
-                            " [platform: %s]\n%s\n%s" %
-                    (ix, default_method, platform_desc, failures.dtype.names,
-                        failures),
+                    err_msg=f"pdf test {ix} failed with method '{default_method}'"
+                            f" [platform: {platform_desc}]\n{failures.dtype.names}\n{failures}",
```
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:5275]: Odd but ok, I guess

**Review by mdhaber — APPROVED** [review]
I skimmed this, mostly trusting that the automated linting didn't make any logical errors. I noted a few of the ugly things I found toward the top, but they are quite common, so I didn't continue, since we are supposed to rush this. Feel free to ignore these suggestions because they're not nearly complete.

~~The one thing I'll ask is why the lint check is passing, because some of these lines are far longer than the limit. Is that check not enabled here?~~

I thought this was related to the line length stuff, but I guess not. We will need a separate pass for that, I suppose.

**Comment by lucascolley** [comment]
Thanks Matt, I added the `needs-work` label since I intend to do one full pass for style myself.

**Comment by lucascolley** [comment]
https://github.com/astral-sh/ruff/issues/8683 at least some of the formatting issues will be fixed in the next ruff release. I might wait for that and remake the PR (as long as a force push wouldn't be too disruptive?)

**Comment by stefanv** [comment]
@lucascolley In the interest of time, consider using ruff from their main branch?

**Comment by lucascolley** [comment]
down to 22 errors which need to be fixed manually. There may be more once I rebase for gh-19529.

**Comment by lucascolley** [comment]
A slight change is needed for `mypy` from CI but I'lll wait until I've rebased on gh-19529 and fixed any additional errors. After that, I'll do a pass for style and check if any of the old comments still apply, then mark as ready for review. If everything looks good then, we may want to condense down to two commits for the purpose of the blame-ignore.

**Comment by lucascolley** [comment]
@j-bowhay @stefanv would either of you like to give this a look? No rush at all but I think it's ready. The safe auto fixes should be fine so perhaps just check the unsafe ones and my manual fixes?

File: scipy/optimize/_zeros_py.py [hunk 14]
```diff
@@ -1148,7 +1146,7 @@ def start(self, f, a, b, args=()):
 
         if np.sign(fb) * np.sign(fa) > 0:
             raise ValueError("f(a) and f(b) must have different signs, but "
-                             "f({:e})={:e}, f({:e})={:e} ".format(a, fa, b, fb))
+                             f"f({a:e})={fa:e}, f({b:e})={fb:e} ")
```
Comment (j-bowhay) [inline scipy/optimize/_zeros_py.py:1149]: Is the alignment here correct?

**Review by j-bowhay — APPROVED** [review]
Changes seem reasonable but might be useful for another pair of eyes as it is easy to miss stuff

File: scipy/optimize/_zeros_py.py (hunk 14 above)
Comment (lucascolley) [reply scipy/optimize/_zeros_py.py:1149]: Ah no, good spot 👍

**Comment by stefanv** [comment]
OK, let's get this in before it goes stale. Do you want to make the .git-blame-ignore-revs commit(s)?

**Comment by lucascolley** [comment]
> OK, let's get this in before it goes stale. Do you want to make the .git-blame-ignore-revs commit(s)?

👍 I'll address Jake's comment, squash down to 1 clean-up commit and add it to the ignored revisions (tomorrow). Thanks all for the quick review!

File: scipy/optimize/_zeros_py.py (hunk 14 above)
Comment (lucascolley) [reply scipy/optimize/_zeros_py.py:1149]: On second thought I think this is correct.

**Comment by lucascolley** [comment]
@stefanv good to go once CI is okay?

**Comment by stefanv** [comment]
I see the same failure on PRs recently merged to main, so let's get this in before it becomes impossible to merge.

**Comment by stefanv** [comment]
Thanks very much @lucascolley.

**Comment by lucascolley** [comment]
thanks all!

I imagine there isn't appetite for any more big changes soon, but if there are any other rules which you think are worth our time @stefanv, let me know and I'm happy to look into them.

**Pull request merged**